src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,395 @@
+
+  void gen_nondet_init(
+    const exprt &expr,
+    const source_locationt &loc);
+};
+
+/*******************************************************************\
+
+Function: symbol_factoryt::allocate_object
+
+ Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+exprt symbol_factoryt::allocate_object(
+  const exprt &target_expr,
+  const typet &allocate_type,
+  const source_locationt &loc,
+  const bool static_lifetime)
+{
+  const typet &allocate_type_resolved=ns.follow(allocate_type);
+  const typet &target_type=ns.follow(target_expr.type().subtype());
+  bool cast_needed=allocate_type_resolved!=target_type;
+  symbolt &aux_symbol=
+    c_declare_new_tmp_symbol(init_code, symbol_table, loc, allocate_type);
+  aux_symbol.is_static_lifetime=static_lifetime;
+
+  exprt object=aux_symbol.symbol_expr();
+  exprt aoe=address_of_exprt(object);
+  if(cast_needed)
+  {
+    aoe=typecast_exprt(aoe, target_expr.type());
+  }
+  code_assignt assign(target_expr, aoe);
+  assign.add_source_location()=loc;
+  init_code.copy_to_operands(assign);
+
+  input_recorder.record_input(aux_symbol.base_name, object);
+
+  return aoe;
+}
+
+/*******************************************************************\
+
+Function: symbol_factoryt::gen_nondet_init
+
+ Inputs:
+  expr -
+  loc -
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void symbol_factoryt::gen_nondet_init(
+  const exprt &expr,
+  const source_locationt &loc)
+{
+  const typet &type=ns.follow(expr.type());
+
+  if(type.id()==ID_pointer)
+  {
+    // dereferenced type
+    const pointer_typet &pointer_type=to_pointer_type(type);
+    const typet &subtype=ns.follow(pointer_type.subtype());
+
+    code_labelt set_null_label;
+    code_labelt init_done_label;
+
+    static size_t synthetic_constructor_count=0;
+
+    if(!assume_non_null)
+    {
+      auto returns_null_sym=c_declare_new_tmp_symbol(
+        init_code,
+        symbol_table,
+        loc,
+        c_bool_typet(1),
+        "pointer_is_null");
+
+      auto returns_null=returns_null_sym.symbol_expr();
+      auto assign_returns_null=
+          code_assignt(returns_null, c_get_nondet_bool(returns_null_sym.type));
+      assign_returns_null.add_source_location()=loc;
+      init_code.move_to_operands(assign_returns_null);
+
+      auto set_null_inst=
+        code_assignt(expr, null_pointer_exprt(pointer_type));
+      set_null_inst.add_source_location()=loc;
+
+      std::string fresh_label=
+        "post_synthetic_malloc_"+std::to_string(++synthetic_constructor_count);
+      set_null_label=code_labelt(fresh_label, set_null_inst);
+
+      input_recorder.record_input(returns_null_sym.base_name, returns_null);
+
+      init_done_label=code_labelt(fresh_label+"_init_done", code_skipt());
+
+      code_ifthenelset null_check;
+      exprt null_return=
+        zero_initializer(returns_null_sym.type, loc, ns, message_handler);
```

Comment:
You should be able to get away with `from_integer(0, returns_null_sym.type)`, at which point you might not even need a `message_handlert` anymore.